---
title: "R Notebook"
output: html_notebook
---

## Set Up the RNA-seq
Firstly, we need to do inital the set up, like download necessary packages and load the library.
limma, Glimma, edgeR packages are important analysis tools.
Homo.sapiens which allowd us to get gene name, gene id, and ensembl.
other packages are part for successful run other regular commands.
```{r set up}
#set up the routine
setwd("~/Desktop/TRGN510/trgn510_fall20/")
#set up the cran mirror and choose the nearestby location
#chooseCRANmirror()
options(repos="https://cran.rstudio.com" )
#install pacakges
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("limma")
BiocManager::install("Glimma")
BiocManager::install("edgeR")
BiocManager::install("Homo.sapiens")
BiocManager::install("GenomicDataCommons")
BiocManager::install("ShortRead")
install.packages("stringr")
library(limma)
library(Glimma)
library(edgeR)
library(Homo.sapiens)
library(GenomicDataCommons)
library(ShortRead)
library(stringr)
```
# #Data packaging
## 4.1 read counts from files
We download 4 groups data form TCGA website, for 10 white_female, 10 white_male,10 black_female, 10 black_male.
Since it is non-convenient to combine 40 files to a list, here is a new concept is DGEList.
DGEList can combine many files into one list with unique Entrez gene ids.
```{r reading}
#read the files
files <- c("3b7cb73f-d672-4d68-b28a-e01b82096ba4.htseq.counts.gz","b01ddd22-5864-4e4a-a463-62831421fe99.htseq.counts.gz","6d158f35-9847-41fd-814e-9bb63cca955f.htseq.counts.gz","2c4641bd-928f-4fa2-9931-3351955c2396.htseq.counts.gz","c258e44b-44c5-43d7-835e-912049ae12c4.htseq.counts.gz","7972a299-f3bb-43a5-b75f-a1cf3c998bc5.htseq.counts.gz","df652972-6474-474a-9745-4b58fff68744.htseq.counts.gz","c839556e-515d-4793-9960-0edd0a71a101.htseq.counts.gz","5be2bd1d-24ac-47c5-9e20-33dea602ed65.htseq.counts.gz","d0918436-f01a-49c3-afba-5c29f4076f6c.htseq.counts.gz",
           "f561e8fc-3ed9-4d8f-be9f-800966251738.htseq.counts.gz","e95d6e6b-6aaf-4003-9dc6-4b464a9d1725.htseq.counts.gz","cd8d48a8-d183-48ab-82f8-ffb11224239a.htseq.counts.gz","ac3af473-48c0-4fb1-9f31-56ba6fa744c1.htseq.counts.gz","ef34e0ee-46f0-4a81-b800-dd50cc2266a1.htseq.counts.gz","022cbd07-3e20-4144-b2ab-8c76adffae73.htseq.counts.gz","a571b078-2572-4b5a-b9be-52ad05085e1c.htseq.counts.gz","f9ea29ab-6953-4646-b367-cff2483b9f46.htseq.counts.gz","8238188f-fb01-4f30-9f3f-6bec98625c63.htseq.counts.gz","10d6ec12-4039-4262-b16d-1bd671b99a57.htseq.counts.gz",
           "631fc20d-00c1-4b13-8520-f25f0388e985.htseq.counts.gz","fee154e0-ff94-402d-91e3-6c3d458bee00.htseq.counts.gz","9dd22540-6908-447a-bc26-2da5522f68b6.htseq.counts.gz","e5ea329f-8042-479c-b8ff-fb752b6b44e0.htseq.counts.gz","156c4f8b-4467-46d0-bc55-085a29377f3e.htseq.counts.gz","a28e1869-665e-43a9-8b05-ee0841b672df.htseq.counts.gz","8458d0dd-5bc6-42e7-a60d-2953cec4ac9d.htseq.counts.gz","90af0faf-3e8c-4741-bf49-f87a5b21d54c.htseq.counts.gz","6011bac0-a278-42b4-bfd9-97f1b618926e.htseq.counts.gz","8eb179a6-a0af-4628-aae2-78800db3e9d8.htseq.counts.gz",
           "c9381f51-ec09-4487-88a4-8e94452e4893.htseq.counts.gz","ac5bc405-b210-4bf3-a4e6-6ff071aa59c7.htseq.counts.gz","3ce57f19-ef4b-4bb9-8530-eb0752975f4b.htseq.counts.gz","bf8dd7a7-37b1-4a81-9554-d63fdf2fe4f1.htseq.counts.gz","a24d07c0-9b91-4538-96d2-f97544f4dbb0.htseq.counts.gz","96717cca-f561-43ee-8fac-74c9aa43c202.htseq.counts.gz","e897fe45-26b8-468a-84fd-e1ad54f840e3.htseq.counts.gz","7e9b15b3-eca8-4dd1-8d57-a404ace995d9.htseq.counts.gz","112c839c-41b0-4864-8af1-3fda2e67e4b1.htseq.counts.gz","9e49e3ff-f780-48db-9349-e59e2112b6de.htseq.counts.gz")
#check the file content
read.delim(files[12], nrow=5,header = FALSE)
#combine files to a matrix of counts, use readDGE
x <- readDGE(files, columns=c(1,2))
class(x)
dim(x)
```
## 4.2 Organising sample information
In order to downstream analysis, sample-level information should relate to count matrix.
So we need to contain the expreimental variabels, both biological and technical, which will effect on expression level
Here we identify the white and black as group, female and male as lane
```{r organising sample}
#grep the rownames
samplenames <- substring(colnames(x), 0, nchar(colnames(x)))
samplenames
#set up the each row name
colnames(x) <- samplenames
#group by the old people and young people
group <- c(rep("white",20),rep("black",20))
x$samples$group <- group
lane <- as.factor(rep(c("female","male","female","male"), c(10,10,10,10)))
x$samples$lane <- lane
x$samples
```
## 4.3 Organising gene annotations
We need to get the gene id form the DGEList, such as ENSG2333838.23.
However, this ENSMBL name contains decimals, it is important to grep the information
before the decimal. After that, we can get the genes information form Homo.sapines package,
which contains all gene information. We use ENSMBL as keytype to get the GENEID and GENENAME.
Finally, we need delete the duplicated gene, make sure each gene only show up for one time.
```{r organising gene annotations}
#find the gene id which should be ensg...
geneid <- rownames(x)
geneid
#grep the ensg without decimal
gene <- str_match(geneid, "(\\w*).*")
geneid <- gene[,2]
#go through Homo pacakge and get the gene id and gene Name
genes <- AnnotationDbi::select(Homo.sapiens, 
                keys=geneid,
                columns=c("GENEID","GENENAME"),
                keytype="ENSEMBL")
head(genes)

#delete duplicated, select one of the chromosomes to represent the gene with duplicate annotation
genes <- genes[!duplicated(genes$ENSEMBL),]
#import genes to DGEList
x$genes <- genes
x
```


## 5.1 Transformations from the raw-scale
For differential expression and related analyses, gene expression is rarely considered at the level of raw counts since libraries sequenced at a greater depth will result in higher counts. Rather, it is common practice to transform raw counts onto a scale that accounts for such library size differences.
Usually, the methods that we used are CPM, log-CPM based on sequence.
We are using `cpm` in the edgeR package to convert raw counts to CPM and log-CPM.
```{r transformations }
#raw counts are converted to CPM and log-CPM values using cpm function in edgeR.
cpm <- cpm(x)
lcpm <- cpm(x, log=TRUE)
#This calculation ensures that any two read counts with identical CPM values will also have identical log-CPM values. 
#a couner of zero for this data maps to a log-cpm value of -4.51 adfter adding the prior count or offset
L <- mean(x$samples$lib.size) * 1e-6
M <- median(x$samples$lib.size) * 1e-6
c(L, M)
summary(lcpm)
```

## 5.2 Removing genes that are lowly expressed
All combined data contains highly expressed data and lowly expressed data. We want to test genes that expressed in one condition but not other condition, sometimes some genes are not expressed on both condition.
We can use `filterByExpr` function in the edgeR package to remove those genes.
If the library sizes had been larger then a lower CPM cutoff would have been chosen, because larger library sizes provide better resolution to explore more genes at lower expression levels. 
And then we can plot the distribution of log-CPM values
```{r removing}
#check how many count in the 40 datas are zero
table(rowSums(x$counts==0)==40)

# an automatic way to filter genes, while keeping as many genes as possible with worthwhile counts.
keep.exprs <- filterByExpr(x, group=group)
x <- x[keep.exprs,, keep.lib.sizes=FALSE]
dim(x)

#Plotting the distribution log-CPM values
#reduce the number of genes based on the criterion
lcpm.cutoff <- log2(10/M + 2/L)
library(RColorBrewer)
nsamples <- ncol(x)
col <- brewer.pal(nsamples, "Paired")
par(mfrow=c(1,2))
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.80), las=2, main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
den <- density(lcpm[,i])
lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col=col, bty="n")
lcpm <- cpm(x, log=TRUE)
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="B. Filtered data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
den <- density(lcpm[,i])
lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col=col, bty="n")
```

## 5.3 Normalising gene expression distributions
When we prepare the sample and did sequencing process, there are external factors will affect the expression of individual samples which are not biological interest. For instance, the first package of samples will have higher expression than the second one. So we need to do a normalisation to make sure that expression distributions of each sample are similar across the entire experiment.
```{r normalising}
#we need to normalising the gene to ensure the result are similar
x <- calcNormFactors(x, method = "TMM")
x$samples$norm.factors

#show the effect of normalization, the data was duplicated and adjust, make first sample reduce 5% of the original, but the second sample should increase 5-times larger.
x2 <- x
x2$samples$norm.factors <- 1
x2$counts[,1] <- ceiling(x2$counts[,1]*0.05)
x2$counts[,2] <- x2$counts[,2]*5

#There are two graphs shows unnormalised data and normalised data
par(mfrow=c(1,2))
lcpm <- cpm(x2, log=TRUE)
boxplot(lcpm, las=2, col=col, main="")
title(main="A. Example: Unnormalised data",ylab="Log-cpm")
x2 <- calcNormFactors(x2)  
x2$samples$norm.factors
lcpm <- cpm(x2, log=TRUE)
boxplot(lcpm, las=2, col=col, main="")
title(main="B. Example: Normalised data",ylab="Log-cpm")
```

## 5.4 Unsupervised clustering of samples
In our opinion, one of the most important exploratory plots to examine for gene expression analyses is the multi-dimensional scaling (MDS) plot. This plot will show us how much the gene are related, using group. when we look at the distribution of the plot, we will noticy that more same genes are together, this variables will be more related to the cancer.
```{r Unsupervised}
#use plotMSD in limma package
lcpm <- cpm(x, log=TRUE)
par(mfrow=c(1,2))
col.group <- c("young","old")[group]
levels(col.group) <-  brewer.pal(nlevels(col.group), "Set1")
col.group <- as.character(col.group)
plotMDS(lcpm, labels=group, col=col.group)
title(main="A. Sample groups")

lcpm <- cpm(x, log=TRUE)
par(mfrow=c(1,2))
col.group <- group
levels(col.group) <-  brewer.pal(nlevels(col.group), "Set1")
col.group <- c("pink","orange")[group]
col.group <- as.character(col.group)
col.lane <- lane
levels(col.lane) <-  brewer.pal(nlevels(col.lane), "Set2")
col.lane <- as.character(col.lane)
plotMDS(lcpm, labels=group, col=col.group)
title(main="A. Sample groups")
plotMDS(lcpm, labels=lane, col=col.lane, dim=c(3,4))
title(main="B. Sequencing lanes")
#draw MSD graph use Glimma package where explored multiple dimensions
glMDSPlot(lcpm, labels=paste(group, lane, sep="_"), 
          groups=x$samples[,c(2,5)],launch = FALSE)
```

## 6.1 Creating a design matrix and contrasts
We want to know how different expression level that these two color groups.
```{r creating}
# a design matrix is set up with both the color populations and sequencing lane (batch) information.
design <- model.matrix(~0+group+lane)
colnames(design) <- gsub("group", "", colnames(design))
design

# make a matrix using makeContrasts funtion in the limma packages, which design a compare group
contr.matrix <- makeContrasts(
   whitevsblack = white-black, 
   levels = colnames(design))
contr.matrix
```


## 6.2 Removing heteroscedascity from count data
In limma, linear modelling is carried out on the log-CPM values which are assumed to be normally distributed and the mean-variance relationship is accommodated using precision weights calculated by the `voom `function.
When operating the DGEList, `voom` will automatic adapt the  library sizes and normalisation factors form `x`, in order to convert raw data to log-CPM.
The graphs show how trend is removed after `voom` precision weights are applied to the data.
```{r remove heter}
par(mfrow=c(1,2))
v <- voom(x, design, plot=TRUE)
v

vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)
efit <- eBayes(vfit)
plotSA(efit, main="Final model: Mean-variance trend")

```

## 6.4 Examining the number of DE genes
We can quick check the different expression levels, with the number of significantly up and down regulated genes, all in one table. 
Significance is defined using an adjusted p-value cutoff that is set at 5% by default. 
```{r examing}
# get the table before
summary(decideTests(efit))

# adjust p value which give different result
tfit <- treat(vfit, lfc=0.1)
dt <- decideTests(tfit)
# check the table again
summary(dt)
# get the common gene which differnt expression
de.common <- which(dt[,1]!=0)
length(de.common)

head(tfit$genes$GENENAME[de.common], n=20)

vennDiagram(dt[,1], circle.col=c("turquoise", "salmon"))

write.fit(tfit, dt, file="results.txt")
```

## 6.5 Examining individual DE genes from top to bottom
The top DE genes can be listed using `topTreat` for results using `treat`.
we are sorting the gene as small to large, and provide the info of each gene
```{r examing de}
white.vs.black <- topTreat(tfit, coef=1, n=Inf)
head(white.vs.black)
```

## 6.6 Useful graphical representations of differential expression results
we want to visualize the result using plots, mean-differenece plots, which display log-FCs form the linear model fit against the average log-CPM values can be generated using the `plotMD` function, with the differentially expressed genes highlighted.
```{r graphical}
plotMD(tfit, column=1, status=dt[,1], main=colnames(tfit)[1], 
       xlim=c(-8,13))
# glMDPlot will more clear and dynamic to show the result.
glMDPlot(tfit, coef=1, status=dt, main=colnames(tfit)[1],
         side.main="ENSEMBL", counts=lcpm, groups=group,launch = FALSE)

#A heatmap is created for the top 100 DE genes (as ranked by adjusted p-value) from the basal versus LP contrast using the heatmap.2 function from the gplots package.
library(gplots)
white.vs.black.topgenes <- white.vs.black$ENSEMBL[1:100]
i <- which(v$genes$ENSEMBL %in% white.vs.black.topgenes)
mycol <- colorpanel(1000,"blue","white","red")
heatmap.2(lcpm[i,], scale="row",
   labRow=v$genes$GENENAME[i], labCol=group, 
   col=mycol, trace="none", density.info="none", 
   margin=c(8,6), lhei=c(2,10), dendrogram="column")
```


